subprojects/soak/src/integTest/groovy/org/gradle/launcher/daemon/DaemonPerformanceMonitoringSoakTest.groovy

```diff
@@ -35,25 +35,39 @@ class DaemonPerformanceMonitoringSoakTest extends DaemonMultiJdkIntegrationTest
         )
     }
 
-    def "when build leaks slowly daemon is eventually expired"() {
+    @Unroll
+    def "when build leaks heap space slowly daemon is eventually expired (maxBuilds=#maxBuilds, heapSize=#heapSize, leakRate=#leakRate"() {
         when:
         setupTenuredHeapLeak(leakRate)
+
         then:
-        daemonIsExpiredEagerly(maxBuilds, heapSize)
+        daemonIsExpiredEagerly(maxBuilds, heapSize, heapSize * 2)
 
         where:
         maxBuilds | heapSize | leakRate
-        45        | "200m"   | 600
-        40        | "1024m"  | 4000
+        45        | 200      | 2000
+        40        | 1024     | 14000
     }
 
-    private boolean daemonIsExpiredEagerly(int maxBuilds, String heapSize) {
+    def "when build leaks metaspace slowly daemon is eventually expired"() {
+        when:
+        setupMetaspaceLeak(leakReate)
+
+        then:
+        daemonIsExpiredEagerly(maxBuilds, metaspaceSize * 2, metaspaceSize)
+
+        where:
+        maxBuilds | metaspaceSize | leakReate
+        80        | 200           | 5
```

Comment:
I'm using a low leak rate (and relatively big leaked classes) here since otherwise the metaspace fragments and you get an OOM even though it reports >50% free. I think this is more realistic anyway, since usually builds leak entire classloaders with several MB worth of metadata, but only one or two of them at a time.